tests/robustness/multiclient_test/framework/harness.go

```diff
@@ -0,0 +1,241 @@
+package framework
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"io/ioutil"
+	"log"
+	"os"
+	"path"
+	"syscall"
+
+	"github.com/kopia/kopia/tests/robustness/engine"
+	"github.com/kopia/kopia/tests/robustness/fiofilewriter"
+	"github.com/kopia/kopia/tests/robustness/snapmeta"
+	"github.com/kopia/kopia/tests/tools/fio"
+	"github.com/kopia/kopia/tests/tools/kopiarunner"
+)
+
+const (
+	dataSubPath     = "robustness-data"
+	metadataSubPath = "robustness-metadata"
+)
+
+var repoPathPrefix = flag.String("repo-path-prefix", "", "Point the robustness tests at this path prefix")
+
+// NewHarness returns a test harness.
+func NewHarness(ctx context.Context) *TestHarness {
+	dataRepoPath := path.Join(*repoPathPrefix, dataSubPath)
+	metadataRepoPath := path.Join(*repoPathPrefix, metadataSubPath)
+
+	th := &TestHarness{}
+	th.init(ctx, dataRepoPath, metadataRepoPath)
+
+	return th
+}
+
+// TestHarness provides a Kopia robustness.Engine.
+type TestHarness struct {
+	dataRepoPath string
+	metaRepoPath string
+
+	baseDirPath string
+	fileWriter  *MultiClientFileWriter
+	snapshotter *MultiClientSnapshotter
+	persister   *snapmeta.KopiaPersister
+	engine      *engine.Engine
+
+	skipTest bool
+}
+
+func (th *TestHarness) init(ctx context.Context, dataRepoPath, metaRepoPath string) {
+	th.dataRepoPath = dataRepoPath
+	th.metaRepoPath = metaRepoPath
+
+	// Override ENGINE_MODE env variable. Multiclient tests can only run in SERVER mode.
+	log.Printf("Setting %s to %s\n", snapmeta.EngineModeEnvKey, snapmeta.EngineModeServer)
+
+	err := os.Setenv(snapmeta.EngineModeEnvKey, snapmeta.EngineModeServer)
+	if err != nil {
+		log.Printf("Error setting ENGINE_MODE to server: %s", err.Error())
+		os.Exit(1)
+	}
+
+	// the initialization state machine is linear and bails out on first failure
+	if th.makeBaseDir() && th.getFileWriter() && th.getSnapshotter() &&
+		th.getPersister() && th.getEngine(ctx) {
+		return // success!
+	}
+
+	err = th.Cleanup(ctx)
+	if err != nil {
+		log.Printf("Error cleaning up the engine: %s\n", err.Error())
+		os.Exit(2)
+	}
+
+	if th.skipTest {
+		os.Exit(0)
+	}
+
+	os.Exit(1)
+}
+
+func (th *TestHarness) makeBaseDir() bool {
+	baseDir, err := ioutil.TempDir("", "engine-data-")
+	if err != nil {
+		log.Println("Error creating temp dir:", err)
+		return false
+	}
+
+	th.baseDirPath = baseDir
+
+	return true
+}
+
+func (th *TestHarness) getFileWriter() bool {
+	fw, err := NewMultiClientFileWriter(
+		func() (FileWriter, error) { return fiofilewriter.New() },
+	)
+	if err != nil {
+		if errors.Is(err, fio.ErrEnvNotSet) {
+			log.Println("Skipping robustness tests because FIO environment is not set")
+
+			th.skipTest = true
+		} else {
+			log.Println("Error creating fio FileWriter:", err)
+		}
+
+		return false
+	}
```

Comment:
The constructor returns nil for the error - why so much logic to handle an impossible situation?

Reply:
Thanks for catching this. The constructor should be checking for the FIO environment variable to be set and returning an error accordingly. I will fix this.

Reply:
I added this environment variable check to the harness level.

Typically, we perform this check during the construction of `fio.FileWriter` and throw the `fio.ErrEnvNotSet` error, but the constructor for the `MultiClientFileWriter` doesn't create a new `fio.FileWriter` at the time of its construction.

I considered an alternate approach of creating and cleaning up a `FileWriter` during the construction of the `MultiClientFileWriter` just to see if any errors are thrown, but it seemed like overkill.